lib/components/user/notification-prefs-pane.js

```diff
@@ -28,85 +36,393 @@ const allowedNotificationChannels = [
+      // (outside of Formik because (Phone)Input does not have a standard onChange event or simple valitity test).
+      newPhoneNumber: ''
+    }
+  }
+
+  _handleEditNumber = () => {
+    this.setState({
+      isEditing: true,
+      isPending: false,
+      newPhoneNumber: ''
+    })
+  }
+
+  _handleNewPhoneNumberChange = newPhoneNumber => {
+    this.setState({
+      newPhoneNumber
+    })
+  }
+
+  _handleCancelEditNumber = () => {
+    this.setState({
+      isEditing: false,
+      isPending: false
+    })
+  }
+
+  _handlePhoneNumberKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the phone number field,
+      // prevent form submission and request the code.
+      e.preventDefault()
+      this._handleRequestCode()
+    }
+  }
+
+  _handleValidationCodeKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the validation code field,
+      // prevent form submission and send the validation code.
+      e.preventDefault()
+      this._handleSubmitCode()
+    }
+  }
+
+  _handleRequestCode = async () => {
+    // Send phone verification request with the entered values.
+    const { onRequestCode, setFieldValue, setFieldTouched } = this.props
+    const { currentNumber, currentNumberVerified, newPhoneNumber } = this.state
+
+    // If user enters the same phone number as their current verified number,
+    // just cancel editing without sending a verification SMS.
+    if (newPhoneNumber === currentNumber && currentNumberVerified) {
+      this._handleCancelEditNumber()
+    } else {
+      // Empty the input box (and disable the Send text button).
+      this.setState({
+        newPhoneNumber: ''
+      })
+
+      const { isPhoneNumberVerified, phoneNumber } = await onRequestCode(newPhoneNumber)
+
+      if (phoneNumber === newPhoneNumber && !isPhoneNumberVerified) {
+        // On success (phoneNumber is updated and verified status is false), update the phone number in component state.
+        this.setState({
+          newPhoneNumber: phoneNumber
+        })
+
+        // Set empty code field in Formik's state and prompt for code.
+        setFieldValue('validationCode', '')
+        setFieldTouched('validationCode', false)
+        this.setState({
+          isEditing: false,
+          isPending: true
+        })
+      } else if (currentNumber) {
+        // On failure, exit the editing state if there is a current number.
+        // (Stay in edit mode if there is no current number.)
+        this._handleCancelEditNumber()
+      }
+    }
+  }
+
+  _handleSubmitCode = async () => {
+    const { onSubmitCode, onVerified, setFieldValue, values } = this.props
+    const { validationCode } = values
+
+    // Clear the code field and disable the verify button.
+    // (The code to send is already captured in const above.)
+    setFieldValue('validationCode', '')
+
+    const result = await onSubmitCode(validationCode)
```

Comment:
Similar comment as in `_handleRequestCode`. Everything after this appears to be the result of application state changes. It might make sense to end this method here and handle application state in the render methods or via https://formik.org/docs/api/withFormik#mappropstovalues-props-props--values.